plugins/out_gcs/gcs.c

```diff
@@ -0,0 +1,2063 @@
+            /* Buffer last chunk in file and lock file to prevent further changes */
+            ret = buffer_chunk(ctx, upload_file, chunk, chunk_size,
+                               event_chunk->tag, flb_sds_len(event_chunk->tag));
+            if (ret < 0) {
+                FLB_OUTPUT_RETURN(FLB_RETRY);
+            }
+            gcs_store_file_lock(upload_file);
+
+            /* Add chunk file to upload queue */
+            ret = add_to_queue(ctx, upload_file,
+                               event_chunk->tag, flb_sds_len(event_chunk->tag));
+            if (ret < 0) {
+                FLB_OUTPUT_RETURN(FLB_ERROR);
+            }
+
+            /* Go through upload queue and return error if something went wrong */
+            gcs_upload_queue(config, ctx);
+            if (ctx->upload_queue_success == FLB_FALSE) {
+                ctx->upload_queue_success = FLB_TRUE;
+                FLB_OUTPUT_RETURN(FLB_ERROR);
+            }
+            FLB_OUTPUT_RETURN(FLB_OK);
+        }
+        else {
+            /* Send upload directly without upload queue */
+            ret = send_upload_request(ctx, chunk, upload_file,
+                                      event_chunk->tag,
+                                      flb_sds_len(event_chunk->tag));
+            if (ret < 0) {
+                FLB_OUTPUT_RETURN(FLB_ERROR);
+            }
+            FLB_OUTPUT_RETURN(ret);
+        }
+    }
+
+    /* Buffer current chunk in filesystem and wait for next chunk from engine */
+    ret = buffer_chunk(ctx, upload_file, chunk, chunk_size,
+                       event_chunk->tag, flb_sds_len(event_chunk->tag));
+    if (ret < 0) {
+        FLB_OUTPUT_RETURN(FLB_RETRY);
+    }
+    FLB_OUTPUT_RETURN(FLB_OK);
+}
+
+static int cb_gcs_exit(void *data, struct flb_config *config)
+{
+    int ret;
+    struct flb_gcs *ctx = data;
+
+    if (!ctx) {
+        return 0;
+    }
+
+    if (gcs_store_has_data(ctx) == FLB_TRUE) {
+        flb_stream_disable_async_mode(&ctx->u->base);
+        flb_plg_info(ctx->ins, "Sending all locally buffered data to GCS");
+        ret = put_all_chunks(ctx);
+        if (ret < 0) {
+            flb_plg_error(ctx->ins, "Could not send all chunks on exit");
+        }
+    }
+
+    gcs_store_exit(ctx);
+    gcs_context_destroy(ctx);
+
+    return 0;
+}
+
+
+/* Configuration properties map */
+static struct flb_config_map config_map[] = {
+    {
+     FLB_CONFIG_MAP_STR, "google_service_credentials", (char *)NULL,
+     0, FLB_TRUE, offsetof(struct flb_gcs, credentials_file),
+     "Set the path for the google service credentials file"
+    },
+    {
+     FLB_CONFIG_MAP_STR, "json_date_format", NULL,
+     0, FLB_FALSE, 0,
+    FBL_PACK_JSON_DATE_FORMAT_DESCRIPTION
+    },
+    {
+     FLB_CONFIG_MAP_STR, "json_date_key", "date",
+     0, FLB_TRUE, offsetof(struct flb_gcs, json_date_key),
+    "Specifies the name of the date field in output."
+    },
+    {
+     FLB_CONFIG_MAP_SIZE, "total_file_size", "100000000",
```

Comment:
Use `100MB` as a default would be good for readability.
Or, using DEFAULT_FILE_SIZE here would be better.